### DISCUSSION TITLE
Regrid irregular data in space and time

### DISCUSSION THREAD

**Discussion by nilsleh** [pangeo-data/xESMF#378] [Q&A]
Hi, I am quiet new to working with xarray data from different satellites and models, so apologies for potential misunderstandings. I am trying to work with Sea Surface Height Altimetry data from the SWOT mission, similarly as described in this [challenge](https://github.com/ocean-data-challenges/2024b_DC_SWOTdenoising).

The single data example looks as follows:
![image](https://github.com/pangeo-data/xESMF/assets/35272119/7132c803-20ee-408c-812f-c31950125e36)

Now, there are a several files that track measurements over different time stamps and spatial locations. A subset of such data is given in this [example zip](https://drive.google.com/file/d/1gFmcdvjSE4vTzNSObSdjJOw1GGsyiFMf/view?usp=sharing).

Essentially, my goal would be to have a data product with a regular grid over lat/lon and time, similarly to the [Sea Level Product on Copernicus](https://cds.climate.copernicus.eu/cdsapp#!/dataset/10.24381/cds.4c328c78?tab=overview) so that it can be matched with other data sources of the following structure:

![Screenshot from 2024-07-01 13-33-35](https://github.com/pangeo-data/xESMF/assets/35272119/b5342966-098f-48d7-8e5c-cd7e0f63d51f)

Thanks in advance for any suggestions or pointers.

**Comment by aulemahal** [comment]
Hi @nilsleh !  I'm don't know of people that have used xESMF for this kind of remapping, but I am only a volunteer maintainer and I do not work with satellite data...

My first interrogation would be about the  `time` component. xESMF will only help you with 2D spatial remapping, so I'm guessing whatever  solution you find, it would be repeated on each time of the output.

Looking at your example data and using the first file, I was able to do this:
```python3
import xesmf as xe
import xarray as xr

ds = xr.open_dataset('Téléchargements/example/example_data/datafile_0.nc', chunks={})
grid_out = xe.util.cf_grid_2d(-60, 60, 1, -80, 80, 1)

reg = xe.Regridder(ds, grid_out, method='bilinear')

v1 = reg(ds.variable_1)
```
So this gives a large rectangle domain with only the thin line of data in the middle. I guess one could do that for all files and then average them together ?


Again, completely ignoring the  `time` component, could ignore spatial relations between points of measure and put them all in a list ? This is what ESMF calls a "locstream". However, this restricts us to using the 'nearest' method only.
```python3
dd = {}
for i in range(4):
    dd[i] = xr.open_dataset(f'Téléchargements/example/example_data/datafile_{i}.nc', chunks={})
ddi = {}
for i in range(4):
    ddi[i] = dd[i].drop_vars('time').stack(pts=['num_lines', 'num_pixels']).drop_vars(['num_lines', 'num_pixels', 'i_num_line', 'i_num_pixel'])

ds = xr.concat(ddi.values(), 'pts')
reg = xe.Regridder(ds, grid_out, method='nearest_s2d', locstream_in=True)
v1 = reg(ds.variable_1)
v1.plot()
```
Well indeed, this doesn't work well as with only 4 files, the "nearest" method gives garbage.

I'm sure other people have already thought of this problem with our common python tools...
